Update VSO docs for v0.3.0 (#23281)
hashicorp/vault: website/content/docs/platform/k8s/vso/sources/vault.mdx

---
layout: docs

> page_title: Vault Secrets Operator

Reviewer: ```suggestion
page_title: Vault secrets operator
```

description: >-

>   The Vault Secrets Operator allows Pods to consume Vault secrets natively from Kubernetes Secrets.

Reviewer: ```suggestion
  The Vault secrets operator allows Pods to consume Vault secrets natively from Kubernetes Secrets.
```

---

> # Vault Secrets Operator

Reviewer: ```suggestion
# Vault secrets operator
```

> The Vault Secrets Operator (VSO) supports Vault as a secret source. That means that VSO is able to
> seamlessly integrate with Vault running on any platform.

Reviewer: ```suggestion
The Vault secrets operator (VSO) supports Vault as a secret source, which
lets you seamlessly integrate VSO with a Vault instance running on any
platform.
```

style correction: write in active voice with declarative language

## Supported Vault platform and version

| Platform                                  | Version |
|-------------------------------------------|---------|
| [Vault Enterprise/Community](/vault/docs) | 1.11+   |
| [HCP Vault Dedicated](/hcp/docs/vault)    | 1.11+   |

## Features

> The following Vault specific features are supported by the Vault Secrets Operator:

Reviewer: ```suggestion
Vault secrets operator supports the following Vault features:
```

style correction: write in the active voice

- Sync from multiple instances of Vault.
- All Vault [secret engines](/vault/docs/secrets) supported.
- TLS/mTLS communications with Vault.

> - Support for all VSO features, including performing a *rollout-restart* upon secret rotation or
> during drift remediation.

Reviewer: ```suggestion
- Support for all VSO features, including rollout-restart during secret rotation and drift remediation.
```

### Supported Vault authentication methods

| Backend                                   | Description                                                                                                 |
|-------------------------------------------|-------------------------------------------------------------------------------------------------------------|
| [Kubernetes](/vault/docs/auth/kubernetes) | Relies on short-lived Kubernetes ServiceAccount tokens for Vault authentication                             |
| [JWT](/vault/docs/auth/jwt)               | Relies on either static JWT tokens or short-lived Kubernetes ServiceAccount tokens for Vault authentication |
| [AppRole](/vault/docs/auth/approle)       | Relies on static AppRole credentials for Vault authentication                                               |
| [AWS](/vault/docs/auth/aws)               | Relies on AWS credentials for Vault authentication                                                          |

## Vault access and custom resource definitions

> The Vault connection and authentication configuration is provided by the `VaultConnection` and `VaultAuth` CRDs. These can be considered as
> foundational Custom Resources that all secret replication type resources will reference.

Reviewer: ```suggestion
`VaultConnection` and `VaultAuth` CRDs provide Vault connection and authentication configuration information for the secretes operator.
Consider `VaultConnection` and `VaultAuth`  as  foundational custom
resources used by all your secret replication type resources.
```

Style correction: write in active voice

### VaultConnection custom resource

> Provides the configuration necessary for the Operator to connect to a single Vault server instance.

Reviewer: ```suggestion
Provides the required configuration details for connecting the secrete operator to a single Vault server instance.
```

Style correction: write in active voice

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultConnection
metadata:
  namespace: vso-example
  name: vault-connection
spec:
  # required configuration
  # address to the Vault server.
  address: http://vault.vault.svc.cluster.local:8200

  # optional configuration
  # HTTP headers to be included in all Vault requests.
  # headers: []
  # TLS server name to use as the SNI host for TLS connections.
  # tlsServerName: ""
  # skip TLS verification for TLS connections to Vault.
  # skipTLSVerify: false
  # the trusted PEM encoded CA certificate chain stored in a Kubernetes Secret
  # caCertSecretRef: ""
```

### VaultAuth custom resource

Provide the configuration necessary for the Operator to authenticate to a single Vault server instance as

> specified in a `VaultConnection` Custom Resource.

Reviewer: ```suggestion
specified in a `VaultConnection` custom resource.
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultAuth
metadata:
  namespace: vso-example
  name: vault-auth
spec:
  # required configuration
  # VaultConnectionRef of the corresponding VaultConnection CustomResource.
  # If no value is specified the Operator will default to the `default` VaultConnection,
  # configured in its own Kubernetes namespace.
  vaultConnectionRef: vault-connection
  # Method to use when authenticating to Vault.
  method: kubernetes
  # Mount to use when authenticating to auth method.
  mount: kubernetes

>   # Kubernetes specific auth configuration, requires that the Method be set to kubernetes.

Reviewer 2: I think it would be helpful to add the field `allowedNamespaces` here in a comment, or otherwise point out the feature since it's new and a fairly large impact on UX.

Author: True, I do link out to the api-reference, which includes more details on that feature. Would that be good enough?

Reviewer 2: Yeah I think that is good enough for now.
I'll create an issue for me to add it to docs a bit more visibly when I get time because I think some chunk of users will want to use it.

Author: Good call. I think it would be good to provide more explicit docs around the way VSO with dependent resource references.

  kubernetes:
    # role to use when authenticating to Vault
    role: example
    # ServiceAccount to use when authenticating to Vault
    # it is recommended to always provide a unique serviceAccount per Pod/application
    serviceAccount: default

  # optional configuration
  # Vault namespace where the auth backend is mounted (requires Vault Enterprise)
  # namespace: ""
  # Params to use when authenticating to Vault
  # params: []
  # HTTP headers to be included in all Vault authentication requests.
  # headers: []
```

## Vault secret custom resource definitions

> Provide the configuration necessary for the Operator to replicate a single Vault Secret to a single Kubernetes Secret.
> Each supported CRD is specialized to a *class* of Vault secret, documented below.

Reviewer: ```suggestion
Configure the secrets operator to replicate a single Vault Secret to a single Kubernetes Secret. Each supported CRD is specialized to a Vault secrets
 **class**.
```

Style correction: write in active voice

Reviewer: Rather than referencing content with "documented below", we should either provide the explicit heading or use an in-doc link to take the reader to the right place

### VaultStaticSecret custom resource

> Provides the configuration necessary for the Operator to synchronize a single Vault *static* Secret to a single Kubernetes Secret.<br />
> Supported secrets engines: [kv-v2](/vault/docs/secrets/kv/kv-v2), [kv-v1](/vault/docs/secrets/kv/kv-v1)

Reviewer: ```suggestion
Configures the secret operator to synchronize a single Vault **static** secret to a single Kubernetes Secret.

Supported secrets engines: [kv-v2](/vault/docs/secrets/kv/kv-v2), [kv-v1](/vault/docs/secrets/kv/kv-v1)
```

##### KV version 1 secret example

> The KV secrets engine's `kvv1` mount path is specified under `spec.mount` of `VaultStaticSecret` custom resource. Please consult [KV Secrets Engine - Version 1 - Setup](/vault/docs/secrets/kv/kv-v1#setup) for configuring KV secrets engine version 1. The following results in a request to `http://127.0.0.1:8200/v1/kvv1/eng/apikey/google` to retrieve the secret.

Reviewer: ```suggestion
Set the KV secrets engine (`kvv1`) mount path with the `spec.mount` parameter of
your `VaultStaticSecret` custom resource. For more advanced KV secrets engine
version 1 configuration options, consult the
[KV Secrets Engine - Version 1 - Setup](/vault/docs/secrets/kv/kv-v1#setup)
guide.

For example, to send requests to `http://127.0.0.1:8200/v1/kvv1/eng/apikey/google`
to retrieve secrets:
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultStaticSecret
metadata:
  namespace: vso-example
  name: vault-static-secret-v1
spec:
  vaultAuthRef: vault-auth
  mount: kvv1
  type: kv-v1
  path: eng/apikey/google
  refreshAfter: 60s
  destination:
    create: true
    name: static-secret1
```

##### KV version 2 secret example

> The KV secrets engine's `kvv2` mount path is specified under `spec.mount` of `VaultStaticSecret` custom resource. Please consult [KV Secrets Engine - Version 2 - Setup](/vault/docs/secrets/kv/kv-v2#setup) for configuring KV secrets engine version 2. The following results in a request to `http://127.0.0.1:8200/v1/kvv2/data/eng/apikey/google?version=2` to retrieve the secret.

Reviewer: ```suggestion
Set the KV secrets engine (`kvv2`) mount path with the `spec.mount` parameter of
your `VaultStaticSecret` custom resource. For more advanced KV secrets engine
version 2 configuration options, consult the
[KV Secrets Engine - Version 2 - Setup](/vault/docs/secrets/kv/kv-v2#setup)
guide.

For example, to send requests to `http://127.0.0.1:8200/v1/kvv2/eng/apikey/google`
to retrieve secrets:
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultStaticSecret
metadata:
  namespace: vso-example
  name: vault-static-secret-v2
spec:
  vaultAuthRef: vault-auth
  mount: kvv2
  type: kv-v2
  path: eng/apikey/google
  version: 2
  refreshAfter: 60s
  destination:
    create: true
    name: static-secret2
```

### VaultPKISecret custom resource

> Provides the configuration necessary for the Operator to synchronize a single Vault *PKI* Secret to a single Kubernetes Secret.<br />
> Supported secrets engines: [pki](/vault/docs/secrets/pki)
> 
> The PKI secrets engine's mount path is specified under `spec.mount` of `VaultPKISecret` custom resource. Please consult [PKI Secrets Engine - Setup and Usage](/vault/docs/secrets/pki/setup) for configuring PKI secrets engine. The following results in a request to `http://127.0.0.1:8200/v1/pki/issue/default` to generate TLS certificates.

Reviewer: ```suggestion
`VaultPKISecret` configures the secret operator to synchronize a single Vault
PKI Secret to a single Kubernetes Secret.

Supported secrets engines:

- [pki](/vault/docs/secrets/pki)

Set the PKI secrets engine mount path with the `spec.mount` parameter of your
`VaultPKISecret` custom resource. For more advanced PKI secrets engine
configuration options, consult the
[PKI Secrets Engine - Setup and Usage](/vault/docs/secrets/pki/setup)
guide.

For example, to send requests to `http://127.0.0.1:8200/v1/pki/issue/default`
to generate TLS certificates:
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultPKISecret
metadata:
  namespace: vso-example
  name: vault-pki
spec:
  vaultAuthRef: vault-auth
  mount: pki
  role: default
  commonName: example.com
  format: pem
  expiryOffset: 1s
  ttl: 60s
  namespace: tenant-1
  destination:
    create: true
    name: pki1
```

### VaultDynamicSecret custom resource

> Provides the configuration necessary for the Operator to synchronize a single Vault *dynamic* Secret to a single Kubernetes Secret.<br />
> Supported secrets engines *non-exhaustive*: [databases](/vault/docs/secrets/databases), [aws](/vault/docs/secrets/aws),
> [azure](/vault/docs/secrets/azure), [gcp](/vault/docs/secrets/gcp), ...

Reviewer: ```suggestion
VaultDynamicSecret configures the secret operator to synchronize a single Vault
dynamic Secret to a single Kubernetes Secret.

Supported secrets engines:

- [databases](/vault/docs/secrets/databases)
- [aws](/vault/docs/secrets/aws)
- [azure](/vault/docs/secrets/azure)
- [gcp](/vault/docs/secrets/gcp)
```

Reviewer: Adding "non-exhaustive" isn't helpful for readers. We should either say "all Vault dynamic secrets engines" or take the time to actually list the explicit engines that are currently supported.

Author: Let's save this for another review since these are related to the refactoring of the current docs.

##### Database secret example

> The database secret engine's `db` mount path is specified under `spec.mount` of `VaultDynamicSecret` custom resource. Please consult [Database Secrets Engine - Setup](/vault/docs/secrets/databases#setup) for configuring the database secrets engine. The following results in a request to `http://127.0.0.1:8200/v1/db/creds/my-postgresql-role` to generate a new credential.

Reviewer: ```suggestion
Set the database secret engine mount path (`db`) with the `spec.mount` of your
`VaultDynamicSecret` custom resource. For more advanced database secrets engine
configuration options, consult the
[Database Secrets Engine - Setup](/vault/docs/secrets/databases#setup) guide.

For example, to send requests to
`http://127.0.0.1:8200/v1/db/creds/my-postgresql-role` to generate a new
credential:
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  namespace: vso-example
  name: vault-dynamic-secret-db
spec:
  vaultAuthRef: vault-auth
  mount: db
  path: creds/my-postgresql-role
  destination:
    create: true
    name: dynamic-db
```

##### AWS secret example

> The AWS secrets engine's `aws` mount path is specified under `spec.mount` of `VaultDynamicSecret` custom resource. Please consult [AWS Secrets Engine - Setup](/vault/docs/secrets/aws#setup) for configuring AWS secrets engine.
> 
> The following results in a request to `http://127.0.0.1:8200/v1/aws/creds/my-iam-role` to generate a new IAM credential.

Reviewer: ```suggestion
Set the AWS secrets engine mount path (`aws`) with the `spec.mount` parameter of
your `VaultDynamicSecret` custom resource. For more advanced AWS secrets engine
configuration options, consult the
[AWS Secrets Engine - Setup](/vault/docs/secrets/aws#setup) guide.

For example, to send requests to `http://127.0.0.1:8200/v1/aws/creds/my-iam-role`
to generate a new IAM credential:
```

Author: This is outside of the scope of this PR, since it comes from current documentation. Can we come back to it after. The VSO release is today, and the doc updates are blocking it.

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  namespace: vso-example
  name: vault-dynamic-secret-aws-iam
spec:
  vaultAuthRef: vault-auth
  mount: aws
  path: creds/my-iam-role
  destination:
    create: true
    name: dynamic-aws-iam
```

> The following results in a request to `http://127.0.0.1:8200/v1/aws/sts/my-sts-role` to generate a new STS credential.

Reviewer: ```suggestion
To send requests to `http://127.0.0.1:8200/v1/aws/sts/my-sts-role` to generate a new STS credential:
```

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  namespace: vso-example
  name: vault-dynamic-secret-aws-sts
spec:
  vaultAuthRef: vault-auth
  mount: aws
  path: sts/my-sts-role
  destination:
    create: true
    name: dynamic-aws-sts
```

##### HCP Vault Secrets Example

```yaml
---
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  namespace: vso-example
  name: vault-dynamic-secret-aws-iam-role
spec:
  vaultAuthRef: vault-auth
  mount: aws
  path: creds/my-iam-role
  destination:
    create: true
    name: dynamic-aws-iam-role
```

@include 'vso/blurb-api-reference.mdx'

## Tutorial

> Refer to the [Vault Secrets Operator on
> Kubernetes](/vault/tutorials/kubernetes/vault-secrets-operator) tutorial to
> learn the end-to-end workflow using the Vault Secrets Operator.

Reviewer: ```suggestion
Refer to the [Vault secrets operator on
Kubernetes](/vault/tutorials/kubernetes/vault-secrets-operator) tutorial to
learn the end-to-end workflow using the Vault secrets operator.
```

Author: The suggestion seems good, except it the product name is not properly capitalized. It should be "Vault Secrets Operator"
